# SNEEP: SNP exploration and functional analysis using epigenetic data

SNEEP is a fast method to identify regulatory SNPs (rSNPs) that modify the binding sites of Transcription Factors (TFs) for large collections of SNPs or SNVs provided by the user. 

## Documentation
The full documentation is available on [Read the Docs](https://sneep.readthedocs.io/en/latest/index.html)

![sneep_logo.png](sneep_logo.png)

## News
26.7.2023. Our paper about a first application with SNEEP got published in [Human Genomics](https://humgenomics.biomedcentral.com/articles/10.1186/s40246-023-00513-4)

## References

For our statistical approach, please cite: 

Nina Baumgarten, Laura Rumpf, Thorsten Kessler, Marcel H. Schulz (2023), **A statistical approach to identify regulatory DNA variations**,
[preprint](https://doi.org/10.1101/2023.01.31.526404)

We used our statistical approach to determine non-coding disease genes in cardiovascular diseasee. For more details see: 

Chaonan Zhu, Nina Baumgarten, Meiqian Wu, Yue Wang, Arka Provo Das, Jaskiran Kaur, Fatemeh Behjati Ardakani, Thanh Thuy Duong, Minh Duc Pham, Maria Duda, Stefanie Dimmeler, Ting Yuan, Marcel H. Schulz, Jaya Krishnan (2023), **CVD-associated SNPs with with regulatory potential reveal novel non-coding disease genes**, Human Genomics 2023 [full text](https://humgenomics.biomedcentral.com/articles/10.1186/s40246-023-00513-4)

We a currently working on a paper that introduces the statistical background and all abilities of our SNEEP pipeline in more detail.


